Refactor the Database Object Permissions guide

ptgott opened the pull request:
Closes #41917

Merge the Database Object Permissions guide into the Database Access RBAC guide for greater discoverability and a clearer division of labor between the two guides.

This change also includes the following edits to make the refactor cleaner, since we can include each troubleshooting step as a separate H3 in the dateabase object permissions H2:

- Remove an unnecessary troubleshooting step: One step indicates that import rules are validated, which is unnecessary to document, since validation errors are self explanatory.
- Instead of mentioning the admin user as a troubleshooting step, add a separate H3 for the admin user and describe the `admin_user` field, which was not mentioned in the original database object permissions guide.

ptgott committed:
Refactor the Database Object Permissions guide

Closes #41917

Merge the Database Object Permissions guide into the Database Access
RBAC guide for greater discoverability and a clearer division of labor
between the two guides.

This change also includes the following edits to make the refactor
cleaner, since we can include each troubleshooting step as a separate H3
in the dateabase object permissions H2:

- Remove an unnecessary troubleshooting step: One step indicates that
  import rules are validated, which is unnecessary to document, since
  validation errors are self explanatory.
- Instead of mentioning the admin user as a troubleshooting step, add a
  separate H3 for the admin user and describe the `admin_user` field,
  which was not mentioned in the original database object permissions
  guide.

Tener commented on docs/pages/database-access/rbac.mdx:
````diff
@@ -138,3 +138,213 @@ spec:
+
+### Database admin user
+
+A database admin user is responsible for granting permissions to end users. You
+must specify a database admin user before using database object import rules. To
+specify a database admin user, add the following to a role that restricts
+database object permissions:
+
+```yaml
+spec:
+  # ...
+  admin_user: "teleport-admin"
+  # ...
+```
+
+In this case, the Teleport Database Service expects to activate a user called
+`teleport-admin` in order to execute object import rules.  Ensure that the admin
+user possesses the necessary permissions to manage users in your database,
+otherwise object import rules might fail, depending on how you have configured
+the database:
+
+```code
+$ tsh db connect postgres-db --db-name postgres --db-user teleport-user
+psql: error: connection to server at "localhost" (::1), port 50800 failed: Connection refused
+	Is the server running on that host and accepting TCP/IP connections?
+connection to server at "localhost" (127.0.0.1), port 50800 failed: your Teleport role requires automatic database user provisioning but an attempt to activate database user "teleport-user" failed due to the following error: ERROR: permission denied for table pg_subscription (SQLSTATE 42501)
+ERROR: exit status 2
+```
````
`admin_user` is a property of the database spec, which may be missed by the user without the type of object apparent from the `yaml` listing.

Tener commented on docs/pages/database-access/rbac.mdx:
````diff
@@ -138,3 +138,213 @@ spec:
     db_names: ["{{internal.db_names}}"]
 ```
+
+## Database object import rules
````
The old intro part is missing entirely, and the name is gone too.

> **Database Access Controls** is a Teleport feature that lets you configure
role-based access controls for database objects. A **database
object** can be a table, view, or stored procedure. With Database Access Controls, you can ensure that users only have permissions to manage the data
they need.
>
> You can define **import rules** that instruct the Teleport Database Service to
apply labels to database objects imported from databases that match labels configured within the import rule. When a user connects to a database, the Database Service selectively grants permissions by checking database object labels against the user's Teleport roles.
>
> The Database Service grants object-level permissions for the duration of a
connection and revokes them automatically when the connection ends.

I think we should still use a separate name for this feature along with a brief intro section.

ptgott replied:
I found it difficult to use the term "Database Access Controls" to describe restrictions on access to database objects, since in the context of this guide, it also seems to refer to the use of roles to restrict access to databases in general. In other words, the "Database Role-Based Access Controls" guide would both describe Database Access Controls, which have to do with database objects, and other role fields that do not.

Would it make sense to include this intro for the entire guide and encompass non-object-related role fields within "Database Access Controls"?

Tener replied:
I think it is worthwhile to highlight this feature in some way; the "Database Access Controls" name was picked by the product team to that end. @r0mant your thoughts on this?

r0mant replied:
I think that Database Access Controls is the overall name for the functionality that includes both "old style" database server based and object based levels of granularity. Agree about merging the guides into one, from a user's standpoint, I don't think we want them to think about those as two different features.

Having a some kind of intro that explains that with Database Access Controls you can control access to both entire database servers as well as individual objects within those database servers sounds like a good idea though.

Tener commented on docs/config.json:
```diff
@@ -1264,10 +1264,6 @@
             {
               "title": "PostgreSQL",
               "slug": "/database-access/auto-user-provisioning/postgres/"
-            },
-            {
-              "title": "Database Access Controls",
-              "slug": "/database-access/auto-user-provisioning/database-access-controls/"
```
Should we add a redirect?

ptgott replied:
We're going to be overhauling the organization of the docs site this/next month, so I am holding off on introducing new redirects. We are close to our redirect limit, and are considering new approaches for handling these.

ptgott committed:
Respond to Tener feedback

- Clarify the placing of the `admin_user` field

ptgott committed:
Restore intro section

Per Tener and r0mant feedback, integrate the introduction from the
Database Access Controls page into the newly merged RBAC guide. Frame
Database Access Controls as encompassing both databases and database
objects.

ptgott commented:
@r0mant @Tener I have incorporated your feedback in [9fb105d](https://github.com/gravitational/teleport/pull/41925/commits/9fb105d0d28ed122d17bb15d26c23c5df401d0ae)

r0mant commented on docs/pages/database-access/rbac.mdx:
```diff
@@ -1,22 +1,39 @@
 
-For a more general description of Teleport roles and examples see [RBAC](../access-controls/introduction.mdx), as
-this section focuses on configuring RBAC for database access.
+- **Databases:** databases enrolled with your Teleport cluster.
```
Nit: Maybe clarify this as "database servers"? Cause logical databases within DBMS are also called "databases".

r0mant commented on docs/pages/database-access/rbac.mdx:
```diff
@@ -1,22 +1,39 @@
-For a more general description of Teleport roles and examples see [RBAC](../access-controls/introduction.mdx), as
-this section focuses on configuring RBAC for database access.
+- **Databases:** databases enrolled with your Teleport cluster.
+- **Database objects:** tables, views, or stored
```
Should we mention that this is only for Postgres for now?

r0mant commented on docs/pages/database-access/rbac.mdx:
````diff
@@ -138,3 +155,235 @@ spec:
     db_names: ["{{internal.db_names}}"]
 ```
+
+## Database object import rules
````
We should add a warning/note somewhere here saying that only Postgres is supported for now.

ptgott replied:
Added to the top of the guide

r0mant reviewed (approved):
lgtm with a couple of notes

ptgott committed:
Respond to r0mant feedback

ptgott committed:
Fix spelling

ptgott committed:
Fix linter issues

ptgott committed:
Merge branch 'master' into paul.gottschling/41917-db
